### PR TITLE
Typename_IN filters behind experimental flag

### LINKED ISSUE #4405: Authorization does not work with array-based filters.
**Describe the bug**
The param coming from the filter `title_IN: ["Matrix"]`  is evaluated as `[ { "0": "M", "1": "a", "2": "t", "3": "r", "4": "i", "5": "x" } ]` rather than: `["Matrix"]` if used inside an authorization rule.

**Type definitions**
```graphql
type Movie {
    title: String
}

type Actor
    @authorization(
        validate: [
            {
                when: [BEFORE]
                operations: [READ]
                where: { node: { actedInConnection_SOME: { node: { title_IN: ["Matrix"] } } } }
            }
        ]
    ) {
    name: String!
    actedIn: [Movie!]! @relationship(type: "ACTED_IN", direction: OUT, properties: "ActedIn")
}

interface ActedIn @relationshipProperties {
    screenTime: Int!
}
```

**To Reproduce**
Run the following query:

```graphql
query actors {
  actors {
    name
  }
}
```

**Cypher**
```cypher
  MATCH (this:Actor)
  WITH *
  WHERE apoc.util.validatePredicate(NOT ($isAuthenticated = true AND size([(this)-[this1:ACTED_IN]->(this0:Movie) WHERE ($param1 IS NOT NULL AND this0.title IN $param1) | 1]) > 0), "@neo4j/graphql/FORBIDDEN", [0])
  RETURN this { .name } AS this
```
**Params**
```json
{
    "isAuthenticated": false,
    "param1": [ { "0": "M", "1": "a", "2": "t", "3": "r", "4": "i", "5": "x" } ],
 }
```

**Expected behavior**
The `Cypher.Param` should be evaluated correctly.

**System (please complete the following information):**

* OS: [macOS]
* Version: [e.g. @neo4j/graphql@4.4.4]
* Node.js version: [14.16.0]

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Description

This PR introduces support to the typename_IN filters that allow a user to use typename as a filter in the predicate, this substitutes the empty `_on: { Movie: {} }` filters used for analogous results.

## Complexity

Complexity: Medium

# Issue

Closes  #4405 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4409.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 31 | 39 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 38 | 53 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 54 | 78 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 36 | 47 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2125 | 2249 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2092 | 2162 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 29 | 33 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 30 | 41 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2956 | 3217 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 39 | 48 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 63 | 83 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 122 | 153 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 83 | 54 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 136 | 193 | 3003 |
|  connections.Connection | 14049 | 14049 | 66 | 62 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 105 | 117 | 4516 |
|  create.SimpleMutation | 7 | 7 | 45 | 64 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 50 | 24 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 214 | 197 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 182 | 239 | 2040 |
|  2871.NestedRelationshipFilter | 23097 | 23049 | 95 | 61 | 4439 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 34 | 79 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 68 | 89 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 31 | 38 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 49 | 55 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 16 | 19 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 33 | 39 | 2174 |
|  query.QueryWhere | 9687 | 9668 | 28 | 34 | 2166 |
|  query.SimpleQueryWithNestedWhere | 9854 | 9835 | 39 | 47 | 2166 |
|  query.Nested | 10095502 | 10095502 | 7047 | 7017 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 7047 | 6939 | 2004000 |
|  query.OrFilterOnRelationships | 41119 | 43682 | 134 | 155 | 1388 |
|  query.OrFilterOnRelationshipsAndNested | 35461 | 36617 | 175 | 199 | 1388 |
|  query.QueryWithNestedIn | 13817 | 14330 | 47 | 60 | 1367 |
|  query.NestedConnectionWhere | 9801 | 9801 | 48 | 59 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9858 | 10056 | 64 | 84 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18752 | 18968 | 113 | 162 | 1586 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 135 | 165 | 1134 |
|  query.Fulltext | 64 | 64 | 28 | 34 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 38 | 43 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 51 | 49 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 71 | 61 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14174 | 14137 | 56 | 72 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14174 | 14137 | 107 | 93 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14886 | 14401 | 85 | 134 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14521 | 14332 | 88 | 118 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 68 | 84 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14471 | 14462 | 90 | 100 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 43 | 46 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 78 | 78 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 60 | 72 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 52 | 66 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 44 | 49 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 255 | 235 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 247 | 243 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 94 | 110 | 2002 |

</details>
        
Old Schema Generation: 23.607s
Schema Generation: 23.532s
Old Subgraph Schema Generation: 26.828s
Subgraph Schema Generation: 27.026s

File: packages/graphql/tests/utils/builders/context-builder.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ export class ContextBuilder extends Builder<Neo4jGraphQLTranslationContext, Neo4
                 isAuthenticated: true,
                 isAuthenticatedParam: new Cypher.Param(true),
             },
+            experimental: true,
```
Comment (angrykoala) [inline packages/graphql/tests/utils/builders/context-builder.ts:53]: This should only apply on the experimental tests

**Review by angrykoala — DISMISSED** [review]

File: packages/graphql/src/translate/authorization/utils/populate-where-params.ts [hunk 2]
```diff
@@ -32,11 +32,12 @@ export function populateWhereParams({
     const parsed: GraphQLWhereArg = {};
 
     Object.entries(where).forEach(([k, v]) => {
-        if (Array.isArray(v)) {
+        if (k == "AND" || k === "OR") {
```
[the remark this answers is not in the document]
Comment (MacondoExpress) [reply packages/graphql/src/translate/authorization/utils/populate-where-params.ts:35]: Nice catch!

File: packages/graphql/src/translate/translate-read.ts [hunk 3]
```diff
@@ -37,7 +37,7 @@ export function translateRead(
     varName = "this"
 ): Cypher.CypherResult {
     const { resolveTree } = context;
-    const operationsTreeFactory = new QueryASTFactory(context.schemaModel);
+    const operationsTreeFactory = new QueryASTFactory(context.schemaModel, context.experimental);
```
[the remark this answers is not in the document]
Comment (MacondoExpress) [reply packages/graphql/src/translate/translate-read.ts:40]: We need the experimental flag as the typename filters cannot be constructed when building the specific concrete entities but instead, they have to be constructed when parsing the interface as you need the interface implementations available. I put it in the Context for parity to other user arguments like features/subscriptions, but I'm not against changing it!

**Pull request merged**